0005-kzen-two-party-ecdsa.md

```diff
@@ -0,0 +1,139 @@
+
+The paper by Yehuda Lindell's [Fast Secure Two-party ECDSA Signing](https://eprint.iacr.org/2017/552.pdf)is a way to generate a joint public key and to generate joint signatures under this public key.
+This is usually called signature *aggregation*.
+There's a video by the Lindell summarizing the paper here: https://www.youtube.com/watch?v=pwc_Ork-1aA
+
+Aggregated signatures are a nice thing to have.
+They look exactly like normal signatures but are produced by multiple parties instead of one.
+Practically, this means you don't need to lock Bitcoin to multiple public keys and check the signature with `OP_CHECKMULTISIG`; you can just lock it to a joint public key using a normal `P2WSH` output (i.e. use `OP_CHECKSIG`).
+
+Since ECDSA signatures doesn't have a nice algebraic structure it is **very** difficult to do signature aggregation.
+In schnorr you can quite literally just add each party's share of the signature together.
+The method we have to use in ECDSA is to move the operations to produce the *s* component of the signature into a [Paillier group](https://en.wikipedia.org/wiki/Paillier_cryptosystem) rather than doing them simply modulo the Secp256k1 curve order.
+Once you're in Paillier land things become much easier but it comes at a cost of boatloads of messages and complexity during the key generation and exchange phase.
+It is so complicated that it's amazing that anyone actually bothered to come up with this.
+
+Once the keys are generated, Paillier is used to sign roughly in the this way:
+
+1. P1 sends the Paillier encryption of their private key to P2 (this is actually done in the keygen phase)
+2. P2 uses Paillier homorphic operations with the encrypted private key (addition and multiplication) to build a new ciphertext which has P2's parts of the signature in it and sends it to P1.
+3. P1 then decrypts it and then pretty much has a valid ECDSA signature.
+
+The code for the signing part is in the main 2pECDSA crate: https://github.com/KZen-networks/multi-party-ecdsa/blob/e5a741bf8dd756b650b35ef8d65f6cecbd4f196a/src/protocols/two_party_ecdsa/lindell_2017/
+
+The hard bit is doing the setup phase which involves quite a few proofs.
+Some of them are in the above crate but some of them are imported from elsewhere.
+Here's a list of the main proofs that are needed to implement the protocol.
+
+## Key Generation
+
+### Generating a shared public key
+
+To produce signatures under a shared public key you first need a shared public key.
+Because of rogue key attacks, you can't just have one party send their public key and the other send one back.
+
+The technique used in the paper (and followed by the implementation) is to do a coin tossing like protocol.
+This ensures that the public key is totally random and cannot be biased by the party that chooses their key second.
+
+Note, The ability to simulate a random coin toss gives this protocol very strong security guarantees. Not doing so, doesn't necessarily mean it would be *insecure*.
```

Comment:
```suggestion
Note, The ability to simulate a random coin toss gives this protocol very strong security guarantees.
Not doing so, doesn't necessarily mean it would be *insecure*.
```

I am not really sure, what value the last sentence adds. Does it mean we don't have to do it?

Reply:
Yeah I removed it. We might not have to do it. AFAIK as long as you protect against rouge key attacks with the NIZK proof you should be fine.

i.e. You could do something like this in the *Knowledge of Secret Key* model  (KOSK) as far as I understand.
- P1 -> sends `p1_public_key`,
- P2 -> sends `p2_public_key, schnorr_nizk(p2_public_key)`

The 3 round coin tossing is there because Lindell is doing a simulation based proof. In particular he is simulating a functionality which chooses a uniformly random group element. So the extra messages are an artifact of the proof technique (at least to some extent).

I'm not an expert on this so of course we should just do what the paper says :)